GitLab Collector is not running

Hi , 

Whenever i run gitlab collector , i get following error .
```
2016-11-17 16:15:38,434 INFO  org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=10.14.34.73:27017, type=STANDALONE, state=CONNECTED, ok=true, version=ServerVersion{versionList=[3, 2, 10]}, minWireVersion=0, maxWireVersion=4, electionId=null, maxDocumentSize=16777216, roundTripTimeNanos=912360}
2016-11-17 16:15:39,316 INFO  o.s.j.e.a.AnnotationMBeanExporter - Registering beans for JMX exposure on startup
2016-11-17 16:15:39,327 INFO  com.capitalone.dashboard.Application - Started Application in 9.02 seconds (JVM running for 11.151)
2016-11-17 16:18:00,000 INFO  c.c.d.collector.CollectorTask - Running Collector: Gitlab
2016-11-17 16:18:00,009 INFO  c.c.d.collector.CollectorTask - -----------------------------------
2016-11-17 16:18:00,010 INFO  c.c.d.collector.CollectorTask - Starting...
2016-11-17 16:18:00,010 INFO  c.c.d.collector.CollectorTask - -----------------------------------
2016-11-17 16:18:00,118 ERROR o.s.s.s.TaskUtils$LoggingErrorHandler - Unexpected error occurred in scheduled task.
java.lang.NullPointerException: null
        at com.capitalone.dashboard.collecteur.DefaultGitlabGitClient.normalizeUrl(DefaultGitlabGitClient.java:239) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collecteur.DefaultGitlabGitClient.getCommits(DefaultGitlabGitClient.java:68) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collecteur.GitlabGitCollectorTask.collect(GitlabGitCollectorTask.java:76) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collector.CollectorTask.run(CollectorTask.java:63) ~[core-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81) [spring-context-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_45]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_45]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_45]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
```


Below is my property file 
```properties
dbname=db
dbusername=db
dbpassword=dbpass
dbhost=10.14.34.73
dbport=27017
dbreplicaset=false
dbhostport=10.14.34.73:27017
gitlab.cron=0 0/3 * * * *
gitlab.branche=master
gitlab.apiToken=BrXVX-EFBtkq3Jkv6
gitlab.saveLog=true
gitlab.server=http://10.14.34.73:8181
```

Hi All , 

Thanks . Per suggestion , i cloned new master code and now getting new error .

Can anybody please suggest me on how do i mention port in .properties file .

**My Current Properties File looks like :**
 cat application.properties
dbname=db
dbusername=db
dbpassword=dbpass
dbhost=10.14.34.73
dbport=27017
dbreplicaset=false
dbhostport=10.14.34.73:27017
gitlab.cron=0 0/3 * * * *
gitlab.branche=master
gitlab.apiToken=BrXVX-EFBtkq3Jkv6JCD
gitlab.saveLog=true
gitlab.server=http://10.14.34.73:8181
gitlab.host=10.14.34.73
gitlab.port=8181
gitlab.selfSignedCertificate=false
gitlab.commitThresholdDays=15

6-11-24 00:30:00,004 INFO  c.c.d.collector.CollectorTask - -----------------------------------
2016-11-24 00:30:00,188 ERROR o.s.s.s.TaskUtils$LoggingErrorHandler - Unexpected error occurred in scheduled task.
org.springframework.web.client.ResourceAccessException: I/O error on GET request for "https://10.14.34.73/api/v3/projects/root%2FHygieia/repository/commits/?ref_name&since=2016-11-24T00:20:00Z&per_page=100":Connect to 10.14.34.73:443 [/10.14.34.73] failed: Connection refused; nested exception is org.apache.http.conn.HttpHostConnectException: Connect to 10.14.34.73:443 [/10.14.34.73] failed: Connection refused
        at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:580) ~[spring-web-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:545) ~[spring-web-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:466) ~[spring-web-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at com.capitalone.dashboard.collecteur.DefaultGitlabGitClient.makeRestCall(DefaultGitlabGitClient.java:92) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collecteur.DefaultGitlabGitClient.getCommits(DefaultGitlabGitClient.java:69) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collecteur.GitlabGitCollectorTask.collect(GitlabGitCollectorTask.java:98) ~[gitlab-git-collector-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at com.capitalone.dashboard.collector.CollectorTask.run(CollectorTask.java:63) ~[core-2.0.4-SNAPSHOT.jar!/:2.0.4-SNAPSHOT]
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81) [spring-context-4.1.7.RELEASE.jar!/:4.1.7.RELEASE]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_101]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_101]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_101]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_101]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_101]
Caused by: org.apache.http.conn.HttpHostConnectException: Connect to 10.14.34.73:443 [/10.14.34.73] failed: Connection refused
        at org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:142) ~[httpclient-4.3.6.jar!/:4.3.6]
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:319) ~[httpclient-4.3.6.jar!/:4.3.6]
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:363) ~[httpclient-4.3.6.jar!/:4.3.6]
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:219) ~[httpclient-4.3.6.jar!/:4.3.6]
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:195) ~[httpclient-4.3.6.jar!/:4.3.6

--- a/collectors/scm/gitlab/README.md
+++ b/collectors/scm/gitlab/README.md
@@ -48,13 +48,19 @@ logging.file=./logs/gitlab.log
 #Collector schedule (required)
 gitlab.cron=0 0/1 * * * *
 
-#Gitlab host (optional)
+#Gitlab host (optional, defaults to "gitlab.com")
 gitlab.host=gitlab.company.com
+
+#Gitlab protocol (optional, defaults to "http")
+gitlab.protocol=http
+
+#Gitlab port (optional, defaults to protocol default port)
+gitlab.port=80
 
 #If your instance of Gitlab is using a self signed certificate, set to true, default is false
 gitlab.selfSignedCertificate=false
 
-#set apiKey to use HTTPS Auth
+#Gitlab API Token (required, token of user the collector will use by default, can be overriden on a per repo basis from the UI. API token provided by Gitlab)
 gitlab.apiToken=
 
 #Maximum number of days to go back in time when fetching commits

--- a/collectors/scm/gitlab/README.md
+++ b/collectors/scm/gitlab/README.md
@@ -18,11 +18,11 @@ java -JAR gitlab-collector.jar
 
 ## application.properties
 
-You will need to provide an **application.properties** file that contains information about how to connect to the Dashboard MongoDB database instance, as well as properties the Github collector requires. See the Spring Boot [documentation](http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-external-config-application-property-files) for information about sourcing this properties file.
+You will need to provide an **application.properties** file that contains information about how to connect to the Dashboard MongoDB database instance, as well as properties the Gitlab collector requires. See the Spring Boot [documentation](http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-external-config-application-property-files) for information about sourcing this properties file.
 
 ### Sample application.properties file
 
-```#Database Name 
+``` 
 # Database Name
 dbname=dashboard
 

--- a/collectors/scm/gitlab/src/main/java/com/capitalone/dashboard/collecteur/GitlabSettings.java
+++ b/collectors/scm/gitlab/src/main/java/com/capitalone/dashboard/collecteur/GitlabSettings.java
@@ -15,7 +15,9 @@ import org.springframework.stereotype.Component;
 @ConfigurationProperties(prefix = "gitlab")
 public class GitlabSettings {
     private String cron;
+    private String protocol;
     private String host;
+    private String port;
     private String apiToken;
 	private int firstRunHistoryDays;
 	private boolean selfSignedCertificate;
@@ -28,7 +30,15 @@ public class GitlabSettings {
         this.cron = cron;
     }
 
-    public String getApiToken() {
+    public String getProtocol() {
+		return protocol;
+	}
+
+	public void setProtocol(String protocol) {
+		this.protocol = protocol;
+	}
+
+	public String getApiToken() {
         return apiToken;
     }
 
@@ -42,6 +52,14 @@ public class GitlabSettings {
 
 	public void setHost(String host) {
 		this.host = host;
+	}
+
+	public String getPort() {
+		return port;
+	}
+
+	public void setPort(String port) {
+		this.port = port;
 	}
 
 	public int getFirstRunHistoryDays() {

--- a/collectors/scm/gitlab/src/main/java/com/capitalone/dashboard/collecteur/GitlabUrlUtility.java
+++ b/collectors/scm/gitlab/src/main/java/com/capitalone/dashboard/collecteur/GitlabUrlUtility.java
@@ -25,7 +25,7 @@ public class GitlabUrlUtility {
 	private GitlabSettings gitlabSettings;
 	
 	private static final String GIT_EXTENSION = ".git";
-	private static final String PROTOCOL = "https";
+	private static final String DEFAULT_PROTOCOL = "http";
     private static final String SEGMENT_API = "/api/v3/projects/";
 	private static final String COMMITS_API = "/repository/commits/";
 	private static final String DATE_QUERY_PARAM_KEY = "since";
@@ -45,13 +45,15 @@ public class GitlabUrlUtility {
             repoUrl = StringUtils.removeEnd(repoUrl, GIT_EXTENSION);
         }
         
+        String protocol = StringUtils.isBlank(gitlabSettings.getProtocol()) ? DEFAULT_PROTOCOL : gitlabSettings.getProtocol();
 		String repoName = getRepoName(repoUrl);
 		String host = getRepoHost();
 		String date = getDateForCommits(repo, firstRun);
 
 		UriComponentsBuilder builder = UriComponentsBuilder.newInstance();
-		URI uri = builder.scheme(PROTOCOL)
+		URI uri = builder.scheme(protocol)
 				.host(host)
+				.port(gitlabSettings.getPort())
 				.path(SEGMENT_API)
 				.path(repoName)
 				.path(COMMITS_API)

--- a/collectors/scm/gitlab/src/test/java/com/capitalone/dashboard/collecteur/GitlabUrlUtilityTest.java
+++ b/collectors/scm/gitlab/src/test/java/com/capitalone/dashboard/collecteur/GitlabUrlUtilityTest.java
@@ -34,7 +34,7 @@ public class GitlabUrlUtilityTest {
 		
 		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
 		
-		assertEquals("https", result.getScheme());
+		assertEquals("http", result.getScheme());
 		assertEquals("gitlab.com", result.getHost());
 		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
 		assertTrue(result.getQuery().contains("ref_name=master"));
@@ -46,6 +46,22 @@ public class GitlabUrlUtilityTest {
 	public void shouldBuildApiUrlWithGitExtension() {
 		when(gitlabRepo.getRepoUrl()).thenReturn("https://domain.org/namespace/Hygieia.git");
 		when(gitlabRepo.getBranch()).thenReturn("master");
+		
+		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
+		
+		assertEquals("http", result.getScheme());
+		assertEquals("gitlab.com", result.getHost());
+		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
+		assertTrue(result.getQuery().contains("ref_name=master"));
+		assertTrue(result.getQuery().contains("per_page=100"));
+		assertTrue(result.getQuery().contains("since="));
+	}
+	
+	@Test
+	public void shouldBuildApiUrlWithCustomProtocol() {
+		when(gitlabRepo.getRepoUrl()).thenReturn("https://domain.org/namespace/Hygieia");
+		when(gitlabRepo.getBranch()).thenReturn("master");
+		when(gitlabSettings.getProtocol()).thenReturn("https");
 		
 		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
 		
@@ -65,9 +81,27 @@ public class GitlabUrlUtilityTest {
 		
 		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
 		
-		assertEquals("https", result.getScheme());
+		assertEquals("http", result.getScheme());
 		assertEquals("customhost.com", result.getHost());
 		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
+		assertTrue(result.getQuery().contains("ref_name=master"));
+		assertTrue(result.getQuery().contains("per_page=100"));
+		assertTrue(result.getQuery().contains("since="));
+	}
+	
+	@Test
+	public void shouldBuildApiUrlWithCustomPort() {
+		when(gitlabRepo.getRepoUrl()).thenReturn("https://domain.org/namespace/Hygieia");
+		when(gitlabRepo.getBranch()).thenReturn("master");
+		when(gitlabSettings.getHost()).thenReturn("customhost.com");
+		when(gitlabSettings.getPort()).thenReturn("443");
+		
+		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
+		
+		assertEquals("http", result.getScheme());
+		assertEquals("customhost.com", result.getHost());
+		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
+		assertEquals(443, result.getPort());
 		assertTrue(result.getQuery().contains("ref_name=master"));
 		assertTrue(result.getQuery().contains("per_page=100"));
 		assertTrue(result.getQuery().contains("since="));
@@ -81,7 +115,7 @@ public class GitlabUrlUtilityTest {
 		
 		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, true, 100);
 		
-		assertEquals("https", result.getScheme());
+		assertEquals("http", result.getScheme());
 		assertEquals("gitlab.com", result.getHost());
 		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
 		assertTrue(result.getQuery().contains("ref_name=master"));
@@ -97,7 +131,7 @@ public class GitlabUrlUtilityTest {
 		
 		URI result  = gitlabUrlUtility.buildApiUrl(gitlabRepo, false, 100);
 		
-		assertEquals("https", result.getScheme());
+		assertEquals("http", result.getScheme());
 		assertEquals("gitlab.com", result.getHost());
 		assertEquals("/api/v3/projects/namespace%2FHygieia/repository/commits/", result.getRawPath());
 		assertTrue(result.getQuery().contains("ref_name=master"));

